### DISCUSSION TITLE
Updating Annotations

### DISCUSSION THREAD

**Discussion by tobiasspiegltec** [pymupdf/PyMuPDF#3577] [Looking for help]
### Description of the bug

I have a script that copies information from annots to create identical ones on other pages.

This works great, however, the annots appear weird when opening the PDF document. It can be seen, that the annots have the correct information, but to show the correct appearance, I need to change any property of the annot (and then change it back)
I think that the issue has something to do with updating, however updating the annot and the page doesn't work.
Is there a solution to this problem?

### How to reproduce the bug

The following methods are used:

the low level information is copied using doc.xref_object(xref)
a "dummy" annot is created using page.add_rect_annot() (or any other annot)
the annot is updated with the information using doc.update_object()

### PyMuPDF version

1.24.2

### Operating system

Windows

### Python version

3.11

**Comment by JorjMcKie** [comment]
This is not a bug / issue but a Discussions item.
Transferring to there ...

**Comment by JorjMcKie** [comment]
More information is needed to develop an opinion here. First of all a (small) PDF example and then the code used to achieve your goal.

**Reply by tobiasspiegltec** [reply]
Hello, sorry for my late reply.

you can download the following zip and extract the three files in a folder. It consists of:
- "test.pdf" a dummy pdf file with some random annots on it
- "test_to.pdf" an empty dummy pdf file
- "copy_annots.py" the python main file.

when you execute the .py file the script opens the first pdf, copies the annotation infos and creates dummy annotations on the second pdf and overwrites those with the annotation infos from the first pdf.

when you then open the "test_to.pdf" you can see all annotations, but some or all have a strange appearence. When you take closer look on them, you see that the annots do indeed have the correct information. When you change any property of the annot, it appearance turns to "normal"

i'm using the PDF-XChange Editor to open the PDFs.

[bug_report.zip](https://github.com/user-attachments/files/15927614/bug_report.zip)

**Comment by JorjMcKie** [comment]
Your approach is bound to fail.
Just copying the PDF object definitions will not lead to full annotation definitions in the target PDF.
If you insist on this approach (for whatever motivation) you should at least add an intermediate step (e.g. via using the MuPDF CLI tool `mutool`) that recreates the annotations' appearance streams.
Even then it would remain to be a highly non-canonical way of doing this.

Why don't you switch to an approach that reads out the annotations from the source and interprets their properties for adding them in the target PDF using the standard API.

**Reply by tobiasspiegltec** [reply]
I don't insist on this approach, I didn't event came up with it. That is just the first way that worked for me, while searching through the internet. 

what is the standard API? Can you show me a code example?

It want to point out, that the important feature I need is to keep the information of the optional content groups that the annots are referred to.

Thank you in advance for your help!

**Comment by JorjMcKie** [comment]
You should check the type of each original annotation.
Then insert an annotation of the same type in the target page and immediately update it with the properties of the original one.

**Reply by JorjMcKie** [reply]
Hm, the "standard API" simply is `page.add_xxxx_annot(...)`, where "xxxx" corresponds to the name of the annotation type that you must determine. E.g. if you see a "FreeText" annotation in the source page, use `page.add_freetext_annot(...)` in the target.

**Reply by tobiasspiegltec** [reply]
Sorry for the late response, have been on a longer vacation .
Okay, I see what you mean.
That's a little more coding but I'll try that.

**Reply by tobiasspiegltec** [reply]
Thank you again for your advice!
